pkg/reconciler/kubernetes/tektonpipeline/tektonpipeline.go

```diff
@@ -96,47 +107,166 @@ func (r *Reconciler) FinalizeKind(ctx context.Context, original *v1alpha1.Tekton
 // converge the two.
 func (r *Reconciler) ReconcileKind(ctx context.Context, tp *v1alpha1.TektonPipeline) pkgreconciler.Event {
 	logger := logging.FromContext(ctx)
-	tp.Status.InitializeConditions()
-	tp.Status.ObservedGeneration = tp.Generation
 
-	logger.Infow("Reconciling TektonPipeline", "status", tp.Status)
+	if !tp.Status.IsInitialized() {
+		tp.Status.InitializeConditions()
+	}
+
+	// TODO: add name validation in webhook
 	if tp.GetName() != common.PipelineResourceName {
 		msg := fmt.Sprintf("Resource ignored, Expected Name: %s, Got Name: %s",
 			common.PipelineResourceName,
 			tp.GetName(),
 		)
 		logger.Error(msg)
-		tp.GetStatus().MarkInstallFailed(msg)
+		tp.Status.MarkNotReady(msg)
 		return nil
 	}
 
+	// Pass the object through defaulting
+	tp.SetDefaults(ctx)
+
+	// TODO: If target namespace is changed then?
+	// deleting everything and recreating again?
+	// configmap as owner reference
+
+	// TODO: handle upgrades
+	// by label upgrades can be handled
+	// configmap as owner reference
+	// delete resources removed in newer version
+
 	if err := r.extension.PreReconcile(ctx, tp); err != nil {
 		return err
 	}
-	stages := common.Stages{
-		r.appendPipelineTarget,
-		r.transform,
-		common.Install,
-		common.CheckDeployments,
+
+	// Save the transformed manifest, In further reconciliation,
+	// do the transformation only if spec is changed
+	// otherwise use the same transformed manifest
+	isChanged, err := IsObjectChanged(tp)
+	if err != nil {
+		tp.Status.MarkNotReady("Failed to compute hash: " + err.Error())
+		return err
 	}
-	manifest := r.manifest.Append()
-	return stages.Execute(ctx, &manifest, tp)
-}
 
-// appendPipelineTarget mutates the passed manifest by appending one
-// appropriate for the passed TektonComponent
-func (r *Reconciler) appendPipelineTarget(ctx context.Context, manifest *mf.Manifest, comp v1alpha1.TektonComponent) error {
-	if err := common.AppendTarget(ctx, manifest, comp); err != nil {
+	if isChanged {
+
+		// Change ready status to false if true before proceeding
+		ready := tp.Status.GetCondition(apis.ConditionReady)
+		if ready != nil && ready.Status == corev1.ConditionTrue {
+			tp.Status.MarkNotReady("spec changed, reconciling")
+			return fmt.Errorf("reconcile again")
+		}
+
+		lastAppliedManifest = r.manifest
+		if err := r.transform(ctx, &lastAppliedManifest, tp); err != nil {
+			tp.Status.MarkNotReady("transformation failed: " + err.Error())
+			return err
```

Comment:
it looks like here, the `lastAppliedManifest`  is stored for the sake of Finalizer. 

In other words, we are using `lastAppliedManifest` to communicate between 2 different reconcile loops.
Let us avoid doing that, as this makes a counter-intutive assumption, uncommon assumption that a piece of state will be recorded in a variable by a previous invocation of the reconciler.